With 2.1.300-preview2-008322, Razor precompilation isn't back-compatible with 2.0.0 applications

Repro:

* Create this trivial project: https://gist.github.com/SteveSandersonMS/0d25e89db93b8092577c3e10d428b533
* Run `dotnet publish -c Release`

Expected: Publish succeeds
Actual: With the preview 2 SDK, it fails with the following error:

```
EXEC : warning CS8021: No value for RuntimeMetadataVersion found. No assembly containing System.Object was found nor was a value for RuntimeMetadataVersion specified through options. [C:\Users\username\Desktop\tmpaspnet\tmpaspnet.csproj]
EXEC(1,11): error CS0246: The type or namespace name 'System' could not be found (are you missing a using directive or an assembly reference?) [C:\Users\username\Desktop\tmpaspnet\tmpaspnet.csproj]
EXEC(1,54): error CS0518: Predefined type 'System.String' is not defined or imported [C:\Users\username\Desktop\tmpaspnet\tmpaspnet.csproj]
C:\Users\username\.nuget\packages\microsoft.aspnetcore.mvc.razor.viewcompilation\2.0.0\build\netstandard2.0\Microsoft.AspNetCore.Mvc.Razor.ViewCompilation.targets(60,5): error MSB3073: The command ""C:\Users\username\.dotnet\x64\dotnet.exe" exec --runtimeconfig "C:\Users\username\Desktop\tmpaspnet\bin\Release\netcoreapp2.0\tmpaspnet.runtimeconfig.json" --depsfile "C:\Users\username\Desktop\tmpaspnet\bin\Release\netcoreapp2.0\tmpaspnet.deps.json" "C:\Users\username\.nuget\packages\microsoft.aspnetcore.mvc.razor.viewcompilation\2.0.0\build\netstandard2.0\Microsoft.AspNetCore.Mvc.Razor.ViewCompilation.dll" @"obj\Release\netcoreapp2.0\microsoft.aspnetcore.mvc.razor.viewcompilation.rsp"" exited with code 1. [C:\Users\username\Desktop\tmpaspnet\tmpaspnet.csproj]
```

Note that the same steps work fine with the 2.1.0-preview1 final SDK (and of course with 2.0.x).

--- a/src/Microsoft.NET.Sdk.Razor/build/netstandard2.0/Sdk.Razor.CurrentVersion.targets
+++ b/src/Microsoft.NET.Sdk.Razor/build/netstandard2.0/Sdk.Razor.CurrentVersion.targets
@@ -166,6 +166,13 @@ Copyright (c) .NET Foundation. All rights reserved.
     <MvcRazorCompileOnPublish Condition="'$(MvcRazorCompileOnPublish)' == ''">true</MvcRazorCompileOnPublish>
   </PropertyGroup>
 
+  <PropertyGroup>
+    <!-- Set PreserveCompilationContext = true, if RazorCompileToolSet = PrecompilationTool since the precompilation tool requires it to work correctly.
+      This is an important to maintain back-compat when the Sdk is used to compile 2.0 applications.
+    -->
+    <PreserveCompilationContext Condition="'$(PreserveCompilationContext)'=='' AND '$(ResolvedRazorCompileToolset)' == 'PrecompilationTool'">true</PreserveCompilationContext>
+  </PropertyGroup>
+
   <!--
     Properties that configure Razor SDK, but need to be defined in targets due to evaluation order.
   -->

--- a/test/Microsoft.AspNetCore.Razor.Design.Test/IntegrationTests/BuildIntegrationTest.cs
+++ b/test/Microsoft.AspNetCore.Razor.Design.Test/IntegrationTests/BuildIntegrationTest.cs
@@ -257,7 +257,7 @@ namespace Microsoft.AspNetCore.Razor.Design.IntegrationTests
 
         [Fact]
         [InitializeTestProject("SimpleMvc")]
-        public async Task Build_WithoutViews_ProducesDepsFileWithotCompiilationContext()
+        public async Task Build_WithoutViews_ProducesDepsFileWithoutCompiilationContext()
         {
             Directory.Delete(Path.Combine(Project.DirectoryPath, "Views"), recursive: true);
             var customDefine = "RazorSdkTest";
@@ -273,8 +273,32 @@ namespace Microsoft.AspNetCore.Razor.Design.IntegrationTests
         }
 
         [Fact]
+        [InitializeTestProject("SimpleMvc")]
+        public async Task Build_WithoutViews_ProducesDepsFileWithCompiilationContext_WhenUsingPrecompilationTool()
+        {
+            Directory.Delete(Path.Combine(Project.DirectoryPath, "Views"), recursive: true);
+            var customDefine = "RazorSdkTest";
+            var result = await DotnetMSBuild("Build", $"/p:DefineConstants={customDefine} /p:MvcRazorCompileOnPublish=true");
+
+            Assert.BuildPassed(result);
+
+            Assert.FileExists(result, OutputPath, "SimpleMvc.deps.json");
+            var depsFilePath = Path.Combine(Project.DirectoryPath, OutputPath, "SimpleMvc.deps.json");
+            var dependencyContext = ReadDependencyContext(depsFilePath);
+
+            // Pick a couple of libraries and ensure they have some compile references
+            var packageReference = dependencyContext.CompileLibraries.First(l => l.Name == "Microsoft.AspNetCore.Html.Abstractions");
+            Assert.NotEmpty(packageReference.Assemblies);
+
+            var projectReference = dependencyContext.CompileLibraries.First(l => l.Name == "SimpleMvc");
+            Assert.NotEmpty(packageReference.Assemblies);
+
+            Assert.Contains(customDefine, dependencyContext.CompilationOptions.Defines);
+        }
+
+        [Fact]
         [InitializeTestProject("ClassLibrary")]
-        public async Task Build_ClassLibrary_DoesNotProduceDepsFileWithCompilationContext()
+        public async Task Build_ClassLibrary_ProducesDepsFileWithoutCompiilationContext()
         {
             var result = await DotnetMSBuild("Build");
 

